### PR TITLE
[JENKINS-21486] Fix plugin dependencies resolution

### PR DESCRIPTION
**Pull request by Vlatombe**
https://issues.jenkins-ci.org/browse/JENKINS-21486

Also fixes https://issues.jenkins-ci.org/browse/JENKINS-28955 which is a consequence.
- Check core version. If the current core is older than required, the plugin isn't loaded.
- Check that dependencies are enabled. A disabled optional dependency
  will not prevent a plugin from loading.
- Check versions of dependencies declared by a plugin before loading it.
  If any dependency (even optional) is older than what is required,
  then the plugin isn't loaded.

This should prevent use cases where a plugin is loaded but one of its
dependencies is too old so that :
- its @Extension annotated classes cannot be loaded, causing the full
  Jenkins to blow up with crapload of exceptions which are tedious to
  investigate to understand the root cause.
- NoSuchMethodError and the likes at runtime even though boot has
  completed.

Version check (for setups where version list is manually crafted but yet
works) can be disabled by starting Jenkins with

-Dhudson.PluginWrapper.dependenciesVersionCheck.enabled=true

Minor fixes done while implementing this change :
- Fix version parsing in PluginWrapper.Dependency
- Dynamic plugin load didn't check for disabled flag

### CODE REVIEW

**Comment by Vlatombe** [comment]
Will add a test shortly

**Comment by Vlatombe** [comment]
@jenkinsci/code-reviewers

File: core/src/main/java/hudson/PluginWrapper.java [hunk 1]
```diff
@@ -29,6 +29,7 @@
 import hudson.model.Api;
 import hudson.model.ModelObject;
 import jenkins.YesNoMaybe;
+import jenkins.model.Configuration;
```
Comment (daniel-beck) [inline core/src/main/java/hudson/PluginWrapper.java]: That doesn't seem to be a great choice for configuration options outside `Jenkins` itself.

File: core/src/main/java/hudson/PluginWrapper.java [hunk 2]
```diff
@@ -524,13 +527,40 @@ public boolean hasLicensesXml() {
...
+                    .append(Util.join(missingDependencies, ", "))
+                    .append(" ").append(plural ? "don't" : "doesn't")
+                    .append(" exist. ");
+        }
+        if (!obsoleteDependencies.isEmpty()) {
+            boolean plural = obsoleteDependencies.size() > 1;
+            messageBuilder.append(plural ? "Dependencies " : "Dependency ")
+                    .append(Util.join(obsoleteDependencies, ", "))
+                    .append(" ").append(plural ? "are" : "is")
+                    .append(" older than required.").append(" To disable the version check, use -D")
```
Comment (daniel-beck) [inline core/src/main/java/hudson/PluginWrapper.java]: Not sure this should be advertised so aggressively. We may well remove support for it at some point.

**Comment by daniel-beck** [comment]
Other than the choice of system property, seems :+1:

**Comment by Vlatombe** [comment]
Updated and rebased

**Comment by batmat** [comment]
:+1:

File: core/src/main/java/hudson/PluginWrapper.java [hunk 3]
```diff
@@ -87,6 +87,7 @@
  */
 @ExportedBean
 public class PluginWrapper implements Comparable<PluginWrapper>, ModelObject {
+    public static final boolean DISABLE_PLUGIN_DEPENDENCIES_VERSION_CHECK = Boolean.getBoolean(PluginWrapper.class.getName()+"." + "disablePluginDependenciesVersionCheck");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/PluginWrapper.java]: NIT: Missing javadoc

**Comment by oleg-nenashev** [comment]
I would definitely like to see a test case, which proves the loading does not blow up on optional dependencies. I'm also aware about bundling prebuilt HPI w/o sources. 

Thanks for implementing it in any case! I think it makes sense to integrate it before 2.0

**Comment by jglick** [comment]
Does this supersede #2001? And probably useful to combine with #2098?

File: core/src/main/java/hudson/PluginWrapper.java (hunk 3 above)
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: Prefer to use positive, not negative, sense on options. Thus, a property which refers to _enabling_ the version check, defaulting to true.

**Comment by jglick** [comment]
Long overdue, but I think you forgot the `active` check from #2001.

**Comment by Vlatombe** [comment]
@jglick : IMO both #2002 (supercedes #2001) and #2098 are complementary to this PR.

**Comment by Vlatombe** [comment]
Note: I didn't change the behaviour w.r.t optional dependencies. If an optional dependency is present but with an obsolete version, the plugin will be loaded anyway.

**Comment by jglick** [comment]
That optional dependency behavior would be a bug. If the dependency is present, depending code will assume it can link against it, and will throw `LinkageError`s otherwise.

**Comment by jglick** [comment]
So :bug: for:

* failing to verify that a dependency is not just installed, but enabled
* failing to verify version of an enabled optional dependency

**Comment by jglick** [comment]
Would love to see this in 2.0 by the way.

**Comment by jglick** [comment]
BTW I think it would be better to link this to the original issue, [JENKINS-21486](https://issues.jenkins-ci.org/browse/JENKINS-21486).

**Comment by daniel-beck** [comment]
@jglick If this gets merged to master before the RC it will be in 2.0.

**Comment by Vlatombe** [comment]
I've added active check, and version check for optional dependencies, however it doesn't change anything regarding classloading : the classloader is created even before PluginWrapper instance is created, and it contains all dependencies including optional ones.

As I don't want to disable the plugin if an optional dependency is obsolete, then with the current state of code this optional dependency cannot be excluded from its classloader easily.

**Comment by Vlatombe** [comment]
PluginManager#installPluginWithObsoleteOptionalDependencyCannotAccessIt fails but would require rework of https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/ClassicPluginStrategy.java#L251 to work

**Comment by jglick** [comment]
> I don't want to disable the plugin if an optional dependency is obsolete

Why not? You should! Consider plugin A 1.0

```java
public abstract class X implements ExtensionPoint {
  public abstract void run();
}
```

and 1.1

```java
public abstract class X implements ExtensionPoint {
  public abstract void run();
  protected void prepare() {}
}
```

and a plugin B declaring an optional dependency on A 1.1:

```java
@Extension(optional=true) public class Y extends X {
  @Override public void run() {prepare();}
}
```

As I understand it, you would allow B to be loaded with A 1.0. This will cause `Y` to be included in the extension list, but `Y.run` will throw `NoSuchMethodError`.

I am not grasping for hypothetical examples. All sorts of plugins have optional dependencies which rely heavily on particular APIs introduced in particular versions of those dependencies. The “optional” part only refers to the fact that the dependee can cleanly turn off the entire block of functionality if the dependency is missing. But if it seems to be present, everything will be included, and it had better not throw errors.

**Comment by Vlatombe** [comment]
@jglick fine, in that case i'll just make it fail ;)

**Comment by Vlatombe** [comment]
@jglick done.

Given the amount of commits I'll squash before submitting this for final review/merge.

File: core/src/main/java/hudson/PluginManager.java [hunk 4]
```diff
@@ -489,7 +489,8 @@ public void dynamicLoad(File arc) throws IOException, InterruptedException, Rest
         // so existing plugins can't be depending on this newly deployed one.
 
         plugins.add(p);
-        activePlugins.add(p);
+        if (p.isActive())
```
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java]: Is it even possible for this to be false?

File: core/src/main/java/hudson/PluginWrapper.java [hunk 5]
```diff
@@ -88,6 +88,12 @@
 @ExportedBean
 public class PluginWrapper implements Comparable<PluginWrapper>, ModelObject {
     /**
+     * A plugin won't be loaded unless his declared dependencies are present and match the required minimal version.
+     * This can be set to false to disable the version check (legacy behaviour)
+     */
+    public static final boolean ENABLE_PLUGIN_DEPENDENCIES_VERSION_CHECK = Boolean.parseBoolean(System.getProperty(PluginWrapper.class.getName()+"." + "dependenciesVersionCheck.enable", "true"));
```
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: Could coalesce string literals.

File: core/src/main/java/hudson/PluginWrapper.java [hunk 6]
```diff
@@ -217,10 +223,10 @@ public Dependency(String s) {
             if(idx==-1)
                 throw new IllegalArgumentException("Illegal dependency specifier "+s);
             this.shortName = s.substring(0,idx);
-            this.version = s.substring(idx+1);
-            
+            String version = s.substring(idx+1);
+
             boolean isOptional = false;
-            String[] osgiProperties = s.split(";");
+            String[] osgiProperties = version.split("[;]");
```
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: Not sure what is going on in this hunk. Was the parsing of optional dependencies previously broken somehow?

File: core/src/main/java/hudson/PluginWrapper.java [hunk 7]
```diff
@@ -524,18 +535,70 @@ public boolean hasLicensesXml() {
...
-
         // add the optional dependencies that exists
         for (Dependency d : optionalDependencies) {
-            if (parent.getPlugin(d.shortName) != null)
-                dependencies.add(d);
+            PluginWrapper dependency = parent.getPlugin(d.shortName);
+            if (dependency != null) {
+                boolean valid = true;
+                if (!dependency.isActive()) {
+                    LOGGER.warning(d + " is disabled. This plugin won't be able to access its classes.");
```
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: The user will not know what “this” plugin is.

Probably this warning is just noise anyway. There is already a warning about optional extensions that get skipped. Nothing is wrong when an optional dependency is disabled, so why print a message? We do not do so when the dependency is not installed at all.

File: core/src/main/java/hudson/PluginManager.java (hunk 4 above)
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java]: I see—only under the artificial conditions of `dynamicLoadAndDisable`.

File: core/src/main/java/hudson/PluginWrapper.java (hunk 6 above)
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: I see that you say you did find such a bug, fixing in 936c0ab.

**Comment by jglick** [comment]
:bee:

File: core/src/main/java/hudson/PluginWrapper.java (hunk 6 above)
Comment (Vlatombe) [inline core/src/main/java/hudson/PluginWrapper.java]: Yes, added tests in PluginWrapperTest to regress it.

File: core/src/main/java/hudson/PluginWrapper.java (hunk 7 above)
Comment (Vlatombe) [inline core/src/main/java/hudson/PluginWrapper.java]: With proper log context, I think it is okay, but of course this is only my opinion.
```
Mar 29, 2016 7:43:13 PM hudson.PluginManager dynamicLoad
INFO: Attempting to dynamic load /Users/vincent/Projects/Opensource/jenkins/test/target/hudson3353537990711553137test/plugins/depender-0.0.2.hpi
Mar 29, 2016 7:43:13 PM hudson.PluginWrapper resolvePluginDependencies
WARNING: dependee (0.0.2) optional is disabled. This plugin won't be able to access its classes.
```

Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: From inside `doDynamicLoad` that is fine, but what about a routine Jenkins startup thereafter?

Comment (Vlatombe) [inline core/src/main/java/hudson/PluginWrapper.java]: Indeed, will remove it.

**Comment by Vlatombe** [comment]
Updated PR title/description to link with JENKINS-21486 as requested by @jglick.

I think this is done.

**Comment by jglick** [comment]
re:bee: (it seems you made some edits after my last bee, though you squashed commits so I cannot see them directly…another reason why I hate destructive history operations)

**Comment by Vlatombe** [comment]
sorry for that. Yes my last commit removes the warning you mentioned previously.

**Comment by Vlatombe** [comment]
@daniel-beck @oleg-nenashev is it okay for you both with latest changes based on @jglick inputs?

**Comment by Vlatombe** [comment]
@reviewbybees done

**Comment by kwhetstone** [comment]
:+1:

**Comment by jglick** [comment]
Oh, one important thing that has been totally forgotten here: parsing `Jenkins-Version` (or, if unset, `Hudson-Version`) from the manifest. If that is newer than `Jenkins.getVersion()`, `resolvePluginDependencies` should fail.

File: core/src/main/java/hudson/PluginWrapper.java [hunk 8]
```diff
@@ -534,6 +549,14 @@ public boolean hasLicensesXml() {
      *             thrown if one or several mandatory dependencies doesn't exists.
      */
     /*package*/ void resolvePluginDependencies() throws IOException {
+        String requiredCoreVersion = getRequiredCoreVersion();
+        if (requiredCoreVersion == null) {
+            throw new IOException(shortName + " doesn't declare required core version.");
```
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: Hmm, not sure if this is even legal, but for the sake of compatibility perhaps this should be demoted to a warning?

File: core/src/main/java/hudson/PluginWrapper.java [hunk 9]
```diff
@@ -534,6 +549,14 @@ public boolean hasLicensesXml() {
...
+        } else {
+            if (Jenkins.getVersion().isOlderThan(new VersionNumber(requiredCoreVersion))) {
```
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: Should be made conditional on `ENABLE_PLUGIN_DEPENDENCIES_VERSION_CHECK` I suppose.

**Comment by jglick** [comment]
Not sure what is going on with the CI build—looks like tests passed and then the JVM just got stuck.

**Comment by Vlatombe** [comment]
Yes, happened twice; don't know why.

I'll fix your latest review items.

File: core/src/main/java/hudson/PluginWrapper.java [hunk 10]
```diff
@@ -88,6 +88,12 @@
 @ExportedBean
 public class PluginWrapper implements Comparable<PluginWrapper>, ModelObject {
     /**
+     * A plugin won't be loaded unless his declared dependencies are present and match the required minimal version.
+     * This can be set to false to disable the version check (legacy behaviour)
+     */
+    public static final boolean ENABLE_PLUGIN_DEPENDENCIES_VERSION_CHECK = Boolean.parseBoolean(System.getProperty(PluginWrapper.class.getName()+"." + "dependenciesVersionCheck.enabled", "true"));
```
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java]: Could be `private`.

**Comment by jglick** [comment]
:bee:

**Comment by jglick** [comment]
:bee:

**Pull request merged**